src/azul/collections.py

```diff
@@ -271,3 +273,94 @@ def to_dict(self):
             k: v.to_dict() if isinstance(v, NestedDict) else v
             for k, v in self.items()
         }
+
+
+class OrderedSet(MutableSet[K]):
+    """
+    A mutable set that maintains insertion order. Unlike similar implementations
+    of the same name that are floating around on the internet, it is not a
+    sequence.
+
+    >>> s = OrderedSet(['b', 'a', 'c', 'b'])
+    >>> s
+    OrderedSet(['b', 'a', 'c'])
+
+    >>> s.discard('a')
+    >>> s
+    OrderedSet(['b', 'c'])
+
+    >>> s.add('a')
+    >>> s
+    OrderedSet(['b', 'c', 'a'])
+    """
+
+    def __init__(self, members: Iterable[K] = (), /) -> None:
+        self.inner: dict[K, None] = dict.fromkeys(members)
+
+    def __repr__(self) -> str:
+        contents = repr(list(self)) if self else ''
+        return f'{type(self).__name__}({contents})'
+
+    def __iter__(self) -> Iterator[K]:
+        return iter(self.inner)
+
+    def __len__(self) -> int:
+        return len(self.inner)
+
+    def __eq__(self, other: Any) -> bool:
+        """
+        >>> OrderedSet() == set()
```

Comment:
Label each test case with the name of the property (commutativity, transitivity, ..) it's trying to prove. Add one for symmetry on a non-empty set. Make s2 a `set`. As discussed at length in PL, transitivity involves three sets.